File: src/common/string.zig
/// string.zig — Generic string classification and inspection helpers
///
/// Functions here operate purely on `[]const u8` slices and have no
/// dependency on guidance-domain types.  Suitable for reuse in any Zig
/// tool that needs lightweight text analysis.
const std = @import("std");

/// Checks if a needle substring exists within the haystack, ignoring case sensitivity.
pub fn containsIgnoreCase(haystack: []const u8, needle: []const u8) bool {
    if (needle.len == 0) return true;
    if (needle.len > haystack.len) return false;
    var i: usize = 0;
    while (i + needle.len <= haystack.len) : (i += 1) {
        if (std.ascii.eqlIgnoreCase(haystack[i .. i + needle.len], needle)) return true;
    }
    return false;
}

fn isAlphanumericBoundary(c: u8) bool {
    return !std.ascii.isAlphanumeric(c);
}

fn isIdentBoundary(c: u8) bool {
    return !std.ascii.isAlphanumeric(c) and c != '_';
}

fn containsWordWithBoundary(haystack: []const u8, needle: []const u8, boundary_fn: *const fn (u8) bool) bool {
    if (needle.len > haystack.len) return false;
    var i: usize = 0;
    while (i + needle.len <= haystack.len) : (i += 1) {
        if (!std.ascii.eqlIgnoreCase(haystack[i .. i + needle.len], needle)) continue;
        const left_ok = i == 0 or boundary_fn(haystack[i - 1]);
        const right_end = i + needle.len;
        const right_ok = right_end >= haystack.len or boundary_fn(haystack[right_end]);
        if (left_ok and right_ok) return true;
    }
    return false;
}

/// Checks if a needle substring exists within the haystack array of bytes.
pub fn containsWord(haystack: []const u8, needle: []const u8) bool {
    return containsWordWithBoundary(haystack, needle, isAlphanumericBoundary);
}

/// Checks if a needle substring exists within the haystack as an identifier (underscore-aware).
pub fn containsIdentWord(haystack: []const u8, needle: []const u8) bool {
    return containsWordWithBoundary(haystack, needle, isIdentBoundary);
}

/// Checks if any keywords exist within the source string slice.
pub fn containsAny(source: []const u8, keywords: []const []const u8) bool {
    for (keywords) |kw| {
        if (containsIgnoreCase(source, kw)) return true;
    }
    return false;
}

/// Checks if any word from keywords appears in the given source string.
pub fn containsAnyWord(source: []const u8, keywords: []const []const u8) bool {
    for (keywords) |kw| {
        if (containsWord(source, kw)) return true;
    }
    return false;
}

/// Checks if the input string contains any of the specified extensions.
pub fn hasExtension(s: []const u8, extensions: []const []const u8) bool {
    const dot = std.mem.lastIndexOfScalar(u8, s, '.') orelse return false;
    const ext = s[dot + 1 ..];
    for (extensions) |known| {
        if (std.mem.eql(u8, ext, known)) return true;
    }
    return false;
}

/// Checks if a given slice of bytes matches a specified extension pattern.
pub fn isPathToken(s: []const u8, extensions: []const []const u8) bool {
    if (s.len < 3) return false;
    return hasExtension(s, extensions) or std.mem.indexOf(u8, s, "/") != null;
}

/// Checks if a token matches a pattern resembling an identifier in Zig.
pub fn looksLikeIdentifier(token: []const u8) bool {
    if (token.len < 2) return false;

    // Check for underscore (snake_case / SCREAMING_SNAKE_CASE)
    for (token) |ch| {
        if (ch == '_') return true;
    }

    // Check for uppercase after first character (camelCase, PascalCase)
    var i: usize = 1;
    while (i < token.len) : (i += 1) {
        if (std.ascii.isUpper(token[i])) return true;
    }

    return false;
}

/// Checks if a given relative path matches expected test patterns, returning true or false.
pub fn isTestPath(rel_path: []const u8) bool {
    const basename = std.fs.path.basename(rel_path);
    // Strip extension to get the stem.
    const ext = std.fs.path.extension(basename);
    const stem = if (ext.len > 0) basename[0 .. basename.len - ext.len] else basename;
    if (std.mem.endsWith(u8, stem, "_test")) return true;
    if (std.mem.startsWith(u8, stem, "test_")) return true;
    if (std.mem.eql(u8, stem, "tests")) return true;
    if (std.mem.indexOf(u8, rel_path, "/test/") != null) return true;
    if (std.mem.indexOf(u8, rel_path, "/tests/") != null) return true;
    if (std.mem.indexOf(u8, rel_path, "\\test\\") != null) return true;
    if (std.mem.indexOf(u8, rel_path, "\\tests\\") != null) return true;
    return false;
}

/// Converts a reference to a null-terminated C string into a Zig array slice.
pub fn skillNameFromRef(ref: []const u8) []const u8 {
    const base = std.fs.path.basename(ref);
    if (std.mem.eql(u8, base, "SKILL.md")) {
        const dir = std.fs.path.dirname(ref) orelse return base;
        return std.fs.path.basename(dir);
    }
    return base;
}

/// Converts a null-terminated C string slice into a Zig array of characters.
pub fn langFromPath(path: []const u8) []const u8 {
    if (std.mem.endsWith(u8, path, ".zig")) return "zig";
    if (std.mem.endsWith(u8, path, ".py")) return "python";
    if (std.mem.endsWith(u8, path, ".rs")) return "rust";
    if (std.mem.endsWith(u8, path, ".ts") or std.mem.endsWith(u8, path, ".tsx")) return "typescript";
    if (std.mem.endsWith(u8, path, ".js")) return "javascript";
    return "text";
}

/// Converts a slice of byte slices into a new slice of byte slices, duplicating each string.
pub fn dupeStrings(allocator: std.mem.Allocator, strs: []const []const u8) ![][]const u8 {
    const result = try allocator.alloc([]const u8, strs.len);
    var n: usize = 0;
    errdefer {
        for (result[0..n]) |s| allocator.free(s);
        allocator.free(result);
    }
    for (strs) |s| {
        result[n] = try allocator.dupe(u8, s);
        n += 1;
    }
    return result;
}

/// Converts a null-terminated string into a Zig-safe slice, handling memory allocation internally.
pub fn dupeString(allocator: std.mem.Allocator, s: []const u8) !?[]const u8 {
    if (s.len == 0) return null;
    return try allocator.dupe(u8, s);
}

/// Converts a null-terminated string into a Zig-safe slice, handling memory allocation and validation.
pub fn dupeStringOpt(allocator: std.mem.Allocator, opt: ?[]const u8) !?[]const u8 {
    const s = opt orelse return null;
    if (s.len == 0) return null;
    return try allocator.dupe(u8, s);
}

// =============================================================================
// Embedding text preprocessing — strip boilerplate from doc comments
// =============================================================================

/// Common boilerplate prefixes in auto-generated or template doc comments.
/// Stripping these before embedding improves discriminative signal: the
/// remaining domain nouns and action verbs cluster more tightly in the
/// embedding space than generic management/ownership language.
const STRIP_PREFIXES = [_][]const u8{
    "Returns the ",     "Return the ",     "Returns a ",   "Returns an ",
    "Initializes the ", "Initialize the ", "Initialize ",  "Creates a ",
    "Create a ",        "Creates an ",     "Create an ",   "Checks if ",
    "Check if ",        "Verifies that ",  "Verify that ", "Ensures that ",
    "Ensure that ",     "Manages ",        "Represents ",  "Defines ",
    "Caller must ",     "Caller should ",  "The caller ",  "Not thread-safe",
    "Thread-safe",
};

/// Removes unnecessary comment bytes from a Zig source file.
pub fn stripBoilerplate(comment: []const u8) []const u8 {
    const trimmed = std.mem.trimLeft(u8, comment, " \t");
    for (STRIP_PREFIXES) |pfx| {
        if (std.mem.startsWith(u8, trimmed, pfx)) {
            return trimmed[pfx.len..];
        }
    }
    return trimmed;
}

/// Checks if a comment slice contains invalid or unexpected characters, returning true for noise.
pub fn isNoisyComment(comment: []const u8) bool {
    if (comment.len < 10) return true; // too short to be meaningful
    var non_ws: usize = 0;
    var noisy: usize = 0;
    for (comment) |ch| {
        if (!std.ascii.isWhitespace(ch)) {
            non_ws += 1;
            if (std.ascii.isDigit(ch) or (!std.ascii.isAlphanumeric(ch) and !std.ascii.isWhitespace(ch))) noisy += 1;
        }
    }
    return non_ws > 0 and (noisy * 100 / non_ws) > 50;
}

// =============================================================================
// NL query normalization — strip interrogative prefixes and stop words
// =============================================================================

/// Common English filler words that add no semantic signal when embedding.
/// Stored as a comptime perfect hash — zero runtime overhead.
pub const STOP_WORDS = std.StaticStringMap(void).initComptime(.{
    .{ "the", {} },  .{ "a", {} },    .{ "an", {} },
    .{ "in", {} },   .{ "on", {} },   .{ "at", {} },
    .{ "to", {} },   .{ "of", {} },   .{ "for", {} },
    .{ "with", {} }, .{ "by", {} },   .{ "is", {} },
    .{ "are", {} },  .{ "was", {} },  .{ "be", {} },
    .{ "do", {} },   .{ "does", {} }, .{ "did", {} },
    .{ "done", {} }, .{ "work", {} }, .{ "works", {} },
    .{ "get", {} },  .{ "use", {} },
});

/// Natural-language interrogative prefixes to strip before embedding a query.
/// Stripping these improves cosine similarity against descriptor-format entries.
const NL_PREFIXES = [_][]const u8{
    "what is ",         "what are ",  "what does ",  "what's ",
    "where is ",        "where are ", "where does ", "where can i find ",
    "how does ",        "how do ",    "how can i ",  "how to ",
    "why does ",        "why is ",    "why are ",    "when does ",
    "when is ",         "which ",     "who ",        "whose ",
    "can you explain ", "explain ",   "describe ",   "tell me about ",
    "show me ",         "find ",      "search for ", "look for ",
    "i need ",          "i want ",    "help me ",
};

// =============================================================================
// Hot-path lowercase helpers — zero-allocation for identifiers ≤ 256 bytes
// =============================================================================

/// Copy `src` into `dst`, lowercasing every byte. Returns the written slice.
/// `dst.len` must be >= `src.len`; the result is clamped to `@min(src.len, dst.len)`.
/// Does NOT allocate. Safe to call in tight loops.
pub fn lowerInto(dst: []u8, src: []const u8) []u8 {
    const len = @min(src.len, dst.len);
    return std.ascii.lowerString(dst[0..len], src[0..len]);
}

/// Removes leading null characters from a UTF-8 string slice.
pub fn stripNlPrefix(query: []const u8) []const u8 {
    // Lowercase a buffer large enough for the longest prefix we check.
    const MAX_PREFIX_LEN = 30;
    var buf: [MAX_PREFIX_LEN]u8 = undefined;
    const check_len = @min(query.len, MAX_PREFIX_LEN);
    const lower = std.ascii.lowerString(buf[0..check_len], query[0..check_len]);
    for (NL_PREFIXES) |prefix| {
        if (prefix.len <= check_len and std.mem.startsWith(u8, lower, prefix)) {
            return query[prefix.len..];
        }
    }
    return query;
}

// =============================================================================
// Tests
// =============================================================================

/// Converts a string into a slugified format by trimming, lowercasing, and replacing spaces with hyphens.
pub fn slugify(allocator: std.mem.Allocator, s: []const u8) ![]const u8 {
    var buf: std.ArrayList(u8) = .{};
    var prev_dash = true;
    for (s) |c| {
        if (std.ascii.isAlphanumeric(c)) {
            try buf.append(allocator, std.ascii.toLower(c));
            prev_dash = false;
        } else if (!prev_dash and buf.items.len > 0) {
            try buf.append(allocator, '-');
            prev_dash = true;
        }
    }
    while (buf.items.len > 0 and buf.items[buf.items.len - 1] == '-') {
        buf.items.len -= 1;
    }
    if (buf.items.len == 0) try buf.appendSlice(allocator, "work-item");
    if (buf.items.len > 40) buf.items.len = 40;
    return buf.toOwnedSlice(allocator);
}

// =============================================================================
// Sentence truncation — for LOD text pyramids
// =============================================================================

/// Truncates text at a sentence boundary within max_chars.
/// Searches backwards for `. `, `! `, `? `, `.\n` within the budget.
/// Falls back to word boundary, then hard cut if no boundary found.
pub fn truncateAtSentence(allocator: std.mem.Allocator, text: []const u8, max_chars: usize) ![]const u8 {
    if (text.len <= max_chars) return allocator.dupe(u8, text);

    const window = text[0..max_chars];
    // Search backwards for a sentence-end token.
    var i: usize = max_chars;
    while (i > max_chars / 2) : (i -= 1) {
        const ch = window[i - 1];
        const next = if (i < max_chars) window[i] else ' ';
        if ((ch == '.' or ch == '!' or ch == '?') and (next == ' ' or next == '\n')) {
            return allocator.dupe(u8, text[0..i]);
        }
    }
    // Fallback: word boundary
    i = max_chars;
    while (i > max_chars / 2) : (i -= 1) {
        if (window[i - 1] == ' ') {
            return allocator.dupe(u8, std.mem.trimRight(u8, text[0 .. i - 1], " \t"));
        }
    }
    // Hard cut if no boundary found.
    return allocator.dupe(u8, text[0..max_chars]);
}

/// Extracts the first line of a doc comment (for skeleton summaries).
/// For `/// Function description` returns `Function description`.
/// For `//! Module description` returns `Module description`.
/// Strips leading `/// ` or `//! ` prefix and trailing newlines.
pub fn firstCommentLine(comment: []const u8) []const u8 {
    if (comment.len == 0) return "";

    // Skip leading whitespace
    var start: usize = 0;
    while (start < comment.len and (comment[start] == ' ' or comment[start] == '\t')) {
        start += 1;
    }

    // Strip /// or //! prefix
    if (start + 3 <= comment.len) {
        if ((comment[start] == '/' and comment[start + 1] == '/' and comment[start + 2] == '/') or
            (comment[start] == '/' and comment[start + 1] == '!' and comment[start + 2] == '/'))
        {
            start += 3;
            // Skip space after ///
            while (start < comment.len and comment[start] == ' ') {
                start += 1;
            }
        }
    }

    // Find end of first line
    var end = start;
    while (end < comment.len and comment[end] != '\n' and comment[end] != '\r') {
        end += 1;
    }

    // Trim trailing whitespace
    while (end > start and (comment[end - 1] == ' ' or comment[end - 1] == '\t')) {
        end -= 1;
    }

    return comment[start..end];
}

File: src/common/src/string.rs
use std::collections::HashSet;

lazy_static::lazy_static! {
    pub static ref STOP_WORDS: HashSet<&'static str> = {
        let mut s = HashSet::new();
        for w in &[
            "the", "a", "an", "is", "are", "was", "were", "be", "been", "being",
            "have", "has", "had", "do", "does", "did", "will", "would", "could",
            "should", "may", "might", "shall", "can", "need", "dare", "ought",
            "used", "this", "that", "these", "those", "i", "you", "he", "she",
            "it", "we", "they", "me", "him", "her", "us", "them", "my", "your",
            "his", "its", "our", "their", "mine", "yours", "hers", "its", "ours",
            "theirs", "and", "but", "or", "nor", "not", "so", "yet", "for",
            "in", "on", "at", "to", "by", "with", "from", "of", "as", "into",
            "through", "during", "before", "after", "above", "below", "between",
            "out", "off", "over", "under", "again", "further", "then", "once",
        ] {
            s.insert(*w);
        }
        s
    };
}

pub fn trim_right<'a>(slice: &'a [u8], pattern: &[u8]) -> &'a [u8] {
    let mut end = slice.len();
    while end > 0 && pattern.contains(&slice[end - 1]) {
        end -= 1;
    }
    &slice[..end]
}

pub fn trim_left<'a>(slice: &'a [u8], pattern: &[u8]) -> &'a [u8] {
    let mut start = 0;
    while start < slice.len() && pattern.contains(&slice[start]) {
        start += 1;
    }
    &slice[start..]
}

fn contains_word_with_boundary(text: &str, word: &str, is_boundary: fn(u8) -> bool) -> bool {
    let lower = text.to_lowercase();
    let lower_word = word.to_lowercase();
    let bytes = lower.as_bytes();
    let wb = lower_word.as_bytes();
    if wb.is_empty() || wb.len() > bytes.len() {
        return false;
    }
    let mut i = 0;
    while i + wb.len() <= bytes.len() {
        if &bytes[i..i + wb.len()] == wb {
            let left_boundary = i == 0 || is_boundary(bytes[i - 1]);
            let right_boundary = i + wb.len() == bytes.len() || is_boundary(bytes[i + wb.len()]);
            if left_boundary && right_boundary {
                return true;
            }
        }
        i += 1;
    }
    false
}

fn is_ident_boundary(c: u8) -> bool {
    !c.is_ascii_alphanumeric()
}

pub fn contains_ident_word(haystack: &str, needle: &str) -> bool {
    contains_word_with_boundary(haystack, needle, is_ident_boundary)
}

pub fn contains_any(text: &str, keywords: &[&str]) -> bool {
    let lower = text.to_lowercase();
    keywords.iter().any(|k| lower.contains(&k.to_lowercase()))
}

pub fn contains_any_word(text: &str, keywords: &[&str]) -> bool {
    keywords.iter().any(|k| contains_word(text, k))
}

pub fn contains_ignore_case(text: &str, pattern: &str) -> bool {
    text.to_lowercase().contains(&pattern.to_lowercase())
}

pub fn contains_word(text: &str, word: &str) -> bool {
    contains_word_with_boundary(text, word, |c| !c.is_ascii_alphanumeric())
}

pub fn first_comment_line(text: &str) -> Option<String> {
    let line = text.lines().next()?;
    let trimmed = line
        .trim()
        .trim_start_matches("///")
        .trim()
        .trim_start_matches("//!")
        .trim();
    if trimmed.is_empty() {
        None
    } else {
        Some(trimmed.to_string())
    }
}

pub fn has_extension(path: &str, ext: &str) -> bool {
    let ext = ext.trim_start_matches('.');
    path.to_lowercase()
        .ends_with(&format!(".{}", ext.to_lowercase()))
}

pub fn looks_like_identifier(s: &str) -> bool {
    if s.is_empty() {
        return false;
    }
    let first = s.chars().next().unwrap();
    if !first.is_ascii_alphabetic() && first != '_' {
        return false;
    }
    s.chars().all(|c| c.is_ascii_alphanumeric() || c == '_')
}

pub fn lower_into<'a>(dst: &'a mut [u8], src: &[u8]) -> &'a [u8] {
    let len = src.len().min(dst.len());
    for i in 0..len {
        dst[i] = src[i].to_ascii_lowercase();
    }
    &dst[..len]
}

pub fn slugify(text: &str) -> String {
    text.to_lowercase()
        .chars()
        .filter(|c| c.is_ascii_alphanumeric() || *c == ' ' || *c == '-' || *c == '_')
        .map(|c| if c == ' ' { '-' } else { c })
        .collect::<String>()
        .trim_matches('-')
        .to_string()
}

pub fn truncate_at_sentence(text: &str, max_chars: usize) -> String {
    if text.len() <= max_chars {
        return text.to_string();
    }
    let truncated = &text[..max_chars];
    if let Some(last_period) = truncated.rfind('.') {
        if last_period > max_chars / 2 {
            return text[..=last_period].to_string();
        }
    }
    truncated.to_string()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn contains_ignore_case_basic() {
        assert!(contains_ignore_case("Hello World", "hello"));
        assert!(contains_ignore_case("Hello World", "WORLD"));
        assert!(!contains_ignore_case("Hello World", "goodbye"));
    }

    #[test]
    fn contains_word_boundary() {
        assert!(contains_word("test builder", "builder"));
        assert!(!contains_word("test builders", "builder"));
    }

    #[test]
    fn first_comment_line_strips_prefix() {
        assert_eq!(
            first_comment_line("/// This is a doc comment\n/// more"),
            Some("This is a doc comment".into())
        );
    }

    #[test]
    fn has_extension_variants() {
        assert!(has_extension("file.zig", "zig"));
        assert!(has_extension("file.ZIG", ".zig"));
        assert!(!has_extension("file.zig", "rs"));
    }

    #[test]
    fn looks_like_identifier_various() {
        assert!(looks_like_identifier("foo"));
        assert!(looks_like_identifier("_private"));
        assert!(!looks_like_identifier("123abc"));
    }

    #[test]
    fn slugify_converts() {
        assert_eq!(slugify("Hello World"), "hello-world");
        assert_eq!(slugify("FooBar"), "foobar");
    }

    #[test]
    fn stop_words_contains_expected() {
        assert!(STOP_WORDS.contains("the"));
        assert!(STOP_WORDS.contains("is"));
        assert!(!STOP_WORDS.contains("zig"));
    }

    #[test]
    fn contains_ident_word_no_false_positive_on_substring() {
        assert!(!contains_ident_word("mystructfield", "struct"));
        assert!(!contains_ident_word("unstructured", "struct"));
        assert!(contains_ident_word("test_struct", "struct"));
    }

    #[test]
    fn truncate_at_sentence_boundary() {
        let text = "Hello world. This is a test. More content.";
        let result = truncate_at_sentence(text, 20);
        assert_eq!(result, "Hello world.");
    }

    #[test]
    fn truncate_at_sentence_within_limit() {
        let text = "Short";
        assert_eq!(truncate_at_sentence(text, 100), "Short");
    }

    #[test]
    fn trim_right_basic() {
        assert_eq!(trim_right(b"hello   ", b" "), b"hello");
    }

    #[test]
    fn trim_right_noop() {
        assert_eq!(trim_right(b"hello", b" "), b"hello");
    }

    #[test]
    fn trim_right_all_matching() {
        assert_eq!(trim_right(b"   ", b" "), b"");
    }

    #[test]
    fn trim_right_pattern_subset() {
        assert_eq!(trim_right(b"hello!?!", b"!?"), b"hello");
    }

    #[test]
    fn trim_left_basic() {
        assert_eq!(trim_left(b"   hello", b" "), b"hello");
    }

    #[test]
    fn trim_left_noop() {
        assert_eq!(trim_left(b"hello", b" "), b"hello");
    }

    #[test]
    fn contains_ident_word_basic() {
        assert!(contains_ident_word("my_struct_field", "struct"));
        assert!(!contains_ident_word("mystructfield", "struct"));
    }

    #[test]
    fn contains_ident_word_underscore_boundary() {
        assert!(contains_ident_word("test_foo_bar", "foo"));
        assert!(!contains_ident_word("testfoobar", "foo"));
    }

    #[test]
    fn lower_into_short_src() {
        let mut buf = [0u8; 16];
        let result = lower_into(&mut buf, b"HELLO");
        assert_eq!(result, b"hello");
    }

    #[test]
    fn lower_into_long_src_truncated() {
        let mut buf = [0u8; 4];
        let result = lower_into(&mut buf, b"HELLO WORLD");
        assert_eq!(result, b"hell");
    }

    #[test]
    fn lower_into_empty_src() {
        let mut buf = [0u8; 4];
        let result = lower_into(&mut buf, b"");
        assert_eq!(result, b"");
    }

    #[test]
    fn contains_any_basic() {
        assert!(contains_any("hello world", &["hello"]));
        assert!(contains_any("hello world", &["world", "foo"]));
        assert!(!contains_any("hello world", &["foo"]));
    }

    #[test]
    fn contains_any_word_basic() {
        assert!(contains_any_word("test builder", &["test", "builder"]));
        assert!(contains_any_word("test builder", &["builder"]));
        assert!(!contains_any_word("test builders", &["builder"]));
    }

    #[test]
    fn truncate_at_sentence_no_period() {
        let text = "This is a long string with no period at all in the first half";
        let result = truncate_at_sentence(text, 20);
        assert_eq!(result.len(), 20);
    }

    #[test]
    fn truncate_at_sentence_period_too_early() {
        let text = "A. very long string that continues past the limit";
        let result = truncate_at_sentence(text, 20);
        assert_eq!(result.len(), 20);
    }

    #[test]
    fn contains_ident_word_empty_needle() {
        assert!(!contains_ident_word("test", ""));
    }

    #[test]
    fn contains_ident_word_needle_longer_than_haystack() {
        assert!(!contains_ident_word("abc", "abcdef"));
    }

    #[test]
    fn looks_like_identifier_empty() {
        assert!(!looks_like_identifier(""));
    }

    #[test]
    fn looks_like_identifier_starts_with_digit() {
        assert!(!looks_like_identifier("123abc"));
    }

    #[test]
    fn slugify_trims_dashes() {
        assert_eq!(slugify("-hello-"), "hello");
    }

    #[test]
    fn first_comment_line_with_notice_prefix() {
        assert_eq!(
            first_comment_line("//! Module level doc\n/// member"),
            Some("Module level doc".into())
        );
    }

    #[test]
    fn first_comment_line_empty_after_strip() {
        assert_eq!(first_comment_line("///"), None);
    }

    #[test]
    fn has_extension_case_sensitivity() {
        assert!(has_extension("file.ZIG", ".zig"));
        assert!(!has_extension("file.rs", ".zig"));
    }

    #[test]
    fn contains_ident_word_boundary_special_chars() {
        assert!(contains_ident_word("foo->bar", "bar"));
        assert!(!contains_ident_word("foobar", "bar"));
    }
}
